### ISSUE TITLE
How to calculate the total latency for different collectives?

### ISSUE THREAD

**Issue by stas00** [NVIDIA/nccl#1049]
I'm trying to **calculate how much of an impact latency has over bandwidth of the link for different nccl collectives**, so that when we the users buy/rent network hardware we could make an informed choice. I'd love your help to complete my math.

So I want to be able to calculate the total latency overhead and time for the bandwidth and get the total and the relative impact.

So let's take a very simple practical example of having an 10B param model doing pytorch DDP training on 4x 8-gpu nodes. Say, the model has 10 layers and after each layer's backward step it sends 1B params in gradients using all-reduce overlapping with the compute of the next layer.

N=32

1. Am I correct that all-reduce using the tree algo will incur about `2*log(N)` transfers (well, more correct it'd `2*(n-1)/n` as a coefficient instead of 2, but let's speak many gpus where it's close to 2).

2. What about the ring algo? is it also close to `2*log(N)` for all-reduce?

So if my math is right in the simple case the total latency to perform a single all-reduce would be `latency*2*log(N)`. For example if latency is 0.5usec, then we get `0.5*2*5` = 5usec per layer and 50usec (0.05 secs) per the whole model of 10 layers.

But then there is the switch with a much larger latency, how do I incorporate it into the math - as some of the operations will be done outside of the switch and others via the switch. So somehow now we need to expand the math over `nic_latency` and `switch_latency`.

and to complete the picture let's say the slowest link's bandwidth is 400Gbps, which means that since we need to send a total payload of 640Gb (2 times 10B params at 32 bits = 640) - so this should take 1.6 secs.

And so if my math is correct the total time will take 1.65 secs (1.6+0.05) - in other words the latency is 32 times smaller than the bandwidth transfer.

But I haven't included the switch latency in here, so I think the math is flowed then.

If this is already documented somewhere I'm all ears.

Thank you!

**Comment by stas00** [comment]
On twitter Indu shared some suggestions at how to approach this:

1. Measure host-to-host latency and ignore the number of switches in the path. In most cloud environments, you won't be able to find the number of switches in path.
2. Even if your buffer is large, NCCL does the communication in much smaller chunks. You need to account for that since latency incurred is `latency * number_of_chunks`. There is a variable to tune this. Don't remember on top of my head. Probably NCCL_BUFFSIZE. I remember other knobs too. Worth diving into the code.
3. Check if all_reduce is really your bottleneck. In many cases, reduce_scatter and all_gather are the bottlenecks due to zero style sharding techniques. This matters because all_reduce has a tree algorithm with lower number of hops but reduce_scatter and all_gather only have ring algorithms.

**Comment by stas00** [comment]
This has never been addressed, why was it closed?

> please don't hesitate to reopen it with a quick update

users are not allowed to re-open issues you closed, so the above is a false suggestion.

**Comment by sjeaugey** [comment]
> users are not allowed to re-open issues you closed, so the above is a false suggestion.

Thanks Stas for letting us know. I was not aware of that github restriction, and quite surprised that Github never addressed it, by at least giving the choice of allowing users to reopen or not.

> This has never been addressed, why was it closed?

We're automatically closing old issues to try to sort out which ones are still current.
Over the last years, and due to our heavy load, we were not able to address all requests and they have accumulated. So we're cleaning up, and asking users to ping us if they want us to keep the request open. I know it can be frustrating, and we hope users understand our goal is to actually try to better address issues by restarting from a cleaner state.

In any case -- reopening.

**Comment by sjeaugey** [comment]
> I'm trying to calculate how much of an impact latency has over bandwidth of the link for different nccl collectives

That's something we could probably write a book about. Not an easy task, but one we try to do in NCCL to predict which protocol/algo to pick.
We even expose that model programmatically through [ncclGroupSimulateEnd](https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/api/group.html#c.ncclGroupSimulateEnd). It's not perfect but gives a decent estimate in most cases.

> Am I correct that all-reduce using the tree algo will incur about 2*log(N) transfers (well, more correct it'd 2*(n-1)/n as a coefficient instead of 2, but let's speak many gpus where it's close to 2).

Considering `N` nodes of `G` GPUs, a network latency of `L` and an NVLink latency of `l`,  the tree latency is actually `2*(G-1)*l + 2*log2(N)*L`.

The algorithm bandwidth of the tree is `network bandwidth / 2` (at long as NVLink is not a bottleneck)

> What about the ring algo? is it also close to 2*log(N) for all-reduce?

Ring actually has a linear latency, hence it is something like: `N*(G-1)*l+N*L`.

The algorithm bandwidth of ring is `N*G / (2*(N*G-1)) * network bandwidth` (converges to 1/2 at scale like for the tree).

> For example if latency is 0.5usec, then we get 0.5*2*5 = 5usec per layer and 50usec (0.05 secs) per the whole model of 10 layers.

In our measurements, the Simple protocol latency for NVLink is around 4 us and 14us for the network.
For small messages we'd use the LL protocol which has a latency of .6 and 5 us respectively.
And if you're wondering about latency vs Bandwidth, you may be in the middle, i.e. where LL128 is used with latencies of 1.2 and 8.5 us respectively.

> But then there is the switch with a much larger latency, how do I incorporate it into the math - as some of the operations will be done outside of the switch and others via the switch. So somehow now we need to expand the math over nic_latency and switch_latency.

Switch latency is usually 150ns per hop on IB so it doesn't make a huge difference. But on Ethernet it can be much higher so I agree that may need to be taken into account.

**Comment by stas00** [comment]
Thank you very much for your detailed reply and correcting my math, Sylvain! I always appreciate your notes.

The purpose of this calculation was to attempt to figure out the optimal gradient reduction bucket size. Since we can reduce more frequently with smaller payload (at slower bw), at which bucket size do we get the best overlap with compute when the network isn't very fast. And I wanted to know the impact of the latency for this caclulation.
